In [19]:
import math

# Confidence intervals

When we work with a dataset, we need to keep in mind that in fact we're working with a **sample** of the real **population**. Therefore, any statistic (mean, sd,...etc) that we can compute from our **sample** it will not have the same value as the statistic of the real **population**. It will be a single **estimation** and it will be off by a centain unknown amount called **sampling error**. Furthermore, if we wait until we have more data or we simply pick another dataset from a different source of data and we compute the same statistic we will get a different value because our **sample** will be different. Thus, in general them mean of our sample will not be the same as the mean of the population:

$$(sample\_mean=\bar{x}) \neq (population\_mean=\mu)$$

$$\mu = \bar{x} + \delta(n)$$

where $\delta$ is the **sampling error** that will be different for each sample.This **sampling error** will depend on the **sample size** $n$. The bigger the sample, the smaller the error.

Knowing this, could be a bit dissapointing as we don't have any way to determine how big is going to be the **sampling error**. To get a better idea of what could be the real **population mean** we could collect several samples and compute their corresponding **sample means**. Each sample mean will be off by an unknown amount and we might think that we are in the sample place where we started. However, we could make use of the **central limit theorem** and compute **the mean of several sample means** to get a more reliable **estimation**. But this strategy takes time and/or money and we might mot be able to implement it.

Then, what else we can do? We can use the **central limit theorem** in our favour. What this theorem says is that the **sampling distribution** (the distribtuion followed by the means of our samples) follows a **normal distribution** regardeless of the original **population distribution**. In other words, if we could be able to collect many samples and to compute the mean of each of sample, when we will plot the distribution followed by these **sample means**, we will find a normal distribution centered on the **population mean**. Therefore, if we substract the population mean from our sample mean, the resulting value will be a random value that will follow a **normal distribution** centered on zero:

$$\bar{x}-\mu = -\delta(n) = 𝒩(0,σ)$$

if we divide both sides of the equation by the population's standard deviation we will have a normal distribution with $\mu=0$ and $\sigma=1$:

$$z = \frac{(\bar{x}-\mu)}{\sigma}= -\frac{\delta(n)}{\sigma} = 𝒩(0,1)$$

However, not allways we will know the population's standard deviation $\sigma$ but we can have a reliable estimation using our sample:

$$\sigma ≈ \hat{\sigma}=\sqrt{ \sum \frac{(x - \bar{x})^2}{n-1} }$$

As we know the $𝒩(0,1)$ distribution we can use it to provide a range of values around our sample mean that very likely will contain the real **population value**. The idea is then to take one sample, compute our sample mean $\bar{x}$, and then provide a range of values around our sample mean $\bar{x}$ that very likely will contain the real population mean. How likely? As much as we want: 90%, 95%, 98%...

In [1]:
# 1: confidence interval
import numpy as np


# create some (fake) data of patient's cholesterol level,
# we'll assume this was obtained through measurements
# np.random.normal(mean, sd, number_of_samples)
np.random.seed(17)

In [2]:
# Getting 100 values randomly selected from a normal distribution with mean=5.1 and sd=1.6
patients = np.random.normal(5.1, 1.6, 100)
#     targeted mean value    ^
#  with a standard dev of         ^
# number of random values to generate  ^

In [3]:
patients

array([ 5.54202542,  2.13259507,  6.09824178,  6.93249806,  6.75950475,
        8.11862229,  4.92128274,  4.52063786,  5.33788007,  4.39954696,
        8.5740112 ,  6.9436964 ,  2.18990026,  4.87912105,  5.96374338,
        2.25954834,  7.20380246,  4.34248312,  3.35243216,  4.69995609,
        3.52832911,  6.75003055,  5.88613404,  4.38536543,  3.80982387,
        5.31002841,  3.15990362,  5.35598537,  3.89164314,  5.65983359,
        6.66406682,  4.8782636 ,  5.26617009,  5.58094566,  6.64912847,
        6.49139815,  6.00845294,  5.84445175,  3.23540307,  1.84240833,
        3.25133873, 10.45225183,  5.30276354,  3.98931377,  5.99227909,
        5.25863456,  6.12068188,  6.22497709,  3.63425095,  3.84237724,
        6.89069087,  3.52656623,  5.49123203,  4.16974442,  5.78730224,
        6.37744318,  4.12388792,  6.99664673,  3.96266559,  3.85019371,
        4.73140324,  5.29231213,  3.83844826,  0.36380381,  3.82707021,
        5.5553316 ,  5.93605965,  5.23368711,  9.45327294,  6.07

In [4]:
# we can find out the sample mean, it is
patients.mean()

5.278415670796427

In [5]:
# we can find a confidence interval for the population based on our sample
import scipy.stats

confidence_level = 0.95
degrees_freedom = len(patients) - 1  
sample_mean = np.mean(patients)
# note that we use the standard error of the sample
# is an estimate of the standard error of the population (which is used in the theoretical formula)
sample_standard_error = scipy.stats.sem(patients) # sem = standard error of the mean = std(mean)/sqrt(samplesize)

confidence_interval = scipy.stats.t.interval(confidence_level,
                                             degrees_freedom,
                                             sample_mean,
                                             sample_standard_error)

In [6]:
print( 'confidence interval is ', confidence_interval, '.' )

confidence interval is  (4.920803322187407, 5.6360280194054475) .


In [7]:
#Using the normal distribution
confidence_interval_normal = scipy.stats.norm.interval(confidence_level,
                                             sample_mean,
                                             sample_standard_error)

In [8]:
print( 'confidence interval is ', confidence_interval_normal, '.' )

confidence interval is  (4.925174396868425, 5.631656944724429) .


But what this mean? It means that if we would take a large number of samples of 100 patients, and we compute the confidence intervals around each sample mean,
about 95% of the confidence intervals will contain the real population mean  within the (lower,upper) values

In [12]:
scipy.stats.norm.ppf(0.05/2)

-1.9599639845400545

In [13]:
scipy.stats.norm.ppf(1-(0.05/2))

1.959963984540054

# Hypothesis testing

Hypothesis tests can be bradly classified into two types:


*   One sided
*   Two sided

The type of test, depends on the alternate Hypothesis. If the alternate hypothesis contains $H_{1 or A}$ an $\neq$, then is a **two sided test** as it contains **two rejection areas**. In any other case, it will be a **one sided test**.

In addition, we can find two sub types of **one sided tests** depending on which side of the distribution is located the rejection area:

* Left side
* Right side

Again, the location of the rejection area depends on the alternate hypothesis $H_{1/a}$. To know where is located we need to look at the unequality. If the alternate hypothesis says $<$m then the rejection area is on the **left hand side of the distribution**. Otherwise in on the **right hand side**.

What we do with a Hypotesis testing is get a sample and based on our sample try to reject the **null hypothesis**. To do this, we compute what is the probability to get a sample as **extreme or more** than the one we have obtained according to the distribution followed by the **null hypotesis**. In other words: when the **null hypothesis is True** what is the probability to get a sample like the one we have obtained? If this probability is **too small** we will **reject the null hypothesis** (even though that it can be possible to get it). Otherwise, we accept the **null hypothesis**. Then, the question is:

What probability we consider a small probability? 0.05, 0.01, 0.001?

This value is what we call our **significance level $\alpha$**. The complementary of $\alpha$, ie $1-\alpha$ is called **confidence level** as is tipically given in %: 90%, 95%, 99%...etc. In other words, the **significance level** $\alpha$ is probability threshold to consider a given probability too small or not. For example, if we set $\alpha=0.05$, then any probability smaller than 0.05 will be considered too small and we will reject the **null hypothesis**. Otherwise, we will accept the **null hypothesis**.

## Two sided test example

Let's assume that  someone thinks (hypothesis) that the mean
cholesterol level in the population is 5.6. Then, he/she takes blood samples from 100 individuals, and computes the sample mean. He/she want to make a hypothesis test. Let's see how to do it using Python.

In this case, the $H_{0}$ hypothesis is:

* $H_{0}: \mu = 5.6$
* $H_{A}: \mu \neq 5.6$

As the alternate hypothesis contains an $\neq$ we have two rejection areas. Why?
Because the sample mean could be different from the reference value because it's bigger or lower than the reference value. In general, we will never have a value equal to the population mean because of the **sampling error**. Every sample is random, therefore, the values in the sample will be random and the sample mean will be a random value. The question is:

How big can be the discrepancy between our sample mean and the reference value in order to reject the $H_{0}$ hypothesis? In other words, how likely/unlikely is to get our sample mean given the $H_{0}$ hypothesis?

In this case, $\alpha$ is the size of the **total rejection area**. However, we have **two** rejection areas equaly size. Therefore, the size of both have to add $\alpha$. For this reason, the size of each rejection area is $\alpha/2$.

To fix things, let's work with a **significance level $\alpha=0.05$**.

### Using Scipy

In [9]:
from scipy.stats import ttest_1samp
# someone thinks (hypothesises) that the mean of
# cholesterol values in the population is 5.6

# we select a value for alpha of 0.05 (p-value threshold)
# Two-sided test:
# Null hypothesis or H0: mean cholesterol value = 5.6
# Alternative hyp or H1: mean cholesterol value != 5.6
stat, pval = ttest_1samp(patients, popmean = 5.6, alternative = "two-sided")
#Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html

In [10]:
stat, pval

(-1.7843149987052263, 0.07743590566997763)

The first value obtained before, is the value of the statistic while the second is the area that corresponds to the statistic. The other value $pval$ is the total area (probability) associated with the statistic at both sides of the distribution. In other words:

$$p_{val} = P(D|H_{0}=True)= scipy.stats.t.cdf(-1.78, ddof=n-1) +( 1 - scipy.stats.t.cdf(1.78, ddof=n-1)) $$

where $n$ is our sample size (100 in our case). Remember that we have two rejection areas!!!

In ourt case our p_value (0.077) is bigger than our significance level $\alpha/2=0.05/2=0.025$. Therefore **we accept the null hypothesis**: the mean cholesterol value is 5.6

### Manual calculation using the statistic method.

In [15]:
# Computing manually our statistic.
# Note the use of ddof=1 in the np.std. We do it because we don't know the population standard deviation a
# and we are estimating the value with a sample.
t = (np.mean(patients) - 5.6)/(np.std(patients, ddof=1)/np.sqrt(len(patients)))
print("The value of the statistic is: ",t)

The value of the statistic is:  -1.7843149987052263


$\sigma ≈ \sqrt{\sum\frac{1}{(n-1)}(x-\bar{x})^{2}}$

We can see that the value that we obtain for the statistis is the same as the one computed earlier with the `ttest_1samp` function. Great!

Now, let's compute what are the values on the x-axis that will correspond to $\alpha/2=0.025$. Those values are called the **critical values**. As we have two rejection areas, we will have **two critical values**. Once we know them, we will be able to compare our statistic against the **critical values**. Let's see how to do it.

In [12]:
lower_critical_value, upper_critical_value = scipy.stats.t.ppf((0.05/2), df=len(patients)-1), scipy.stats.t.ppf(1-(0.05/2), df=len(patients)-1)
print("The lower critical value is {:.3f}".format(lower_critical_value))
print("The upper critical value is {:.3f}".format(upper_critical_value))

The lower critical value is -1.984
The upper critical value is 1.984


Now, we compare our statistc againt these critical values. If our statistic is within thee two values, we will accept the **null hypothesis**. Otherwise we will reject it. In our case

$$-1.984 < -1.784 < 1.984 $$

Therefore, **we accept the null hypothesis**, which is the same conclussion that we got using the `ttest_1samp` function. Great!

### Manual calculation using the p_value

When we use the p_value, what we need to do, is to compute what is the area that corresponds to our statistic -1.784. This is given by the **cumulative distribution function cdf**. Then, we will have to determine is this area is bigger than $\alpha$ and smaller than $1-\alpha$. If that's the case, we will **accept the null hypothesis**. Let's do it!

In [16]:
print(f'The value of the t statistic is {t}')

The value of the t statistic is -1.7843149987052263


In [17]:
scipy.stats.t.cdf(t, df=len(patients)-1)

0.038717952834988814

In our case:

$$ \alpha/2 = 0.025 < 0.0387 < 1 - \alpha/2 = 0.975 $$

Again, we reach the same conclussion as with the other two methods: **we accept the null hypothesis**. Great!

## One sided test

Now let's assume that our null hypothesis $H_{0}$  is that the cholesterol values are bigger than 5.6. Then in this case the hypothesis are:

* $H_{0}: \mu \geq 5.6$
* $H_{1}: \mu< 5.6$

In this case, we have a one-sided test because we only have one rejection area. Why? Because the only way to have a discrepancy with the null $H_{0}$ hypothesis is to have a sample mean lower that the reference value. In other words, if our sample mean is greater than the reference value, our sample mean value will be consistent with the reference value.

According to the distribution, having a value lower than the reference value is also possible. Then the question is:

How likely is to have a sample mean value lower than the reference value according to the distribution? If this probability is too low (less than $\alpha$), we will reject the null hypothesis $H_{0}$

### Using Scipy

In [14]:
# One-sided test:
# Null hypothesis or H0: mean cholesterol value >= 5.6
# Alternative hyp or H1: mean cholesterol value < 5.6


# ttest_1samp(sample, population_mean)
stat, pval = ttest_1samp(patients, popmean = 5.6, alternative = "less")

In [15]:
stat, pval

(-1.7843149987052263, 0.038717952834988814)

Of course our statistic and the p_value is the same. What has changed are our **null and alternate hypothesis**. Now, we only have **one rejection area** located on the **left hand side of the distribution**. According to the output:

$$0.039 <\alpha$$

Therefore, **we reject the null hypothesis**.

###Manual calculation using the statistic method.

In [16]:
# Computing manually our statistic.
t = (np.mean(patients) - 5.6)/(np.std(patients, ddof=1)/np.sqrt(len(patients)))

Now, as we have only one rejection area, we will only have one **critical value**. Let's compute what is the value on the x-axis that corresponds to an area of $\alpha$.

In [17]:
critical_value = scipy.stats.t.ppf(0.05, df=len(patients)-1)
print("The critical value is {:.3f}".format(critical_value))

The critical value is -1.660


In our case, we have:

$$ -1.784 < -1.660$$

our statistics t is smaller than the critical value. Therefore, we **reject the null hypothesis** which is the same conclussion that we have obtained using the `ttest_1samp` function. Great!!

### Manual calculation using the p_value

To use this method, we need to know which area corresponds to our statistic t and compare it with $alpha$. If the area that corresponds to our statistic is smaller than $\alpha=0.05$ we will **reject the null hypothesis**. Othewise, we will accept it.

In [18]:
scipy.stats.t.cdf(t, df=len(patients)-1)

0.038717952834988814

In our case:

$$ 0.039 < \alpha=0.05$$

Therefore, we **reject the null hypothesis** which is the same conclussion that we got with the other two methods. Great!!

## One sided test (right)

Now le'ts say that we believe that the cholesterol level is smaller than 5.6

Then in this case the hypothesis are:

* $H_{0}: \mu \leq 5.6$
* $H_{1/a}: \mu > 5.6$

In this case, we have a one-sided test because we only have one rejection area. Why? Because the only way to have a discrepancy with the null $H_{0}$ hypothesis is to have a sample mean bigger that the reference value. If our sample mean is lower than the reference value, our sample mean value will be consistent with the reference value.

According to the distribtion, having a value bigger than the reference value is also possible. Then the question is:

How likely is to have a sample mean value bigger than the reference value according to the distribution? If this probability is too low, we will reject the null hypothesis $H_{0}$

### Using Scipy

In [19]:
# One-sided test:
# Null hypothesis or H0: mean cholesterol value <= 5.6
# Alternative hyp or H1: mean cholesterol value > 5.6


# ttest_1samp(sample, population_mean)
stat, pval = ttest_1samp(patients, popmean = 5.6, alternative = "greater")

In [20]:
stat, pval

(-1.7843149987052263, 0.9612820471650112)

Once again, our statistic is the same. However, our **null and alternate hypothesis are different**.

In this case the p_value is bigger than $\alpha=0.05$. Therefore we **accept the null hypothesis**

### Manual calculation using the statistic method.

Again, in this case we will only have one rejection area located on the right hand side of the distribution. The size of it it will be $\alpha$. We need to compare our statistic against the **critical value**. If our statistic is smaller than the critical value, we will **accept the nul hypothesis**. Otherwise, we will **reject the null hypothesis**. Let's compute the **critical value**

In [21]:
# Compute your p_value
critical_value = scipy.stats.t.ppf(1-0.05, df=len(patients)-1)
print("The critical value is {:.3f}".format(critical_value))

The critical value is 1.660


In our case, we have:

$$-1.784 < 1.660$$

Therefore, we **accept the null hypothesis** which is the same conclussion that we obtained with the previous method. Great!!

### Manual calculation using the p_value

Finally, in this case we need to know what is the area that corresponds to our statistic t and compare it against $1 - \alpha$. If our area is smaller than $1 - \alpha$ we **accept the null hypothesis**.

Then, we need to know use the cdf.

In [22]:
scipy.stats.t.cdf(t,df=len(patients)-1)

0.038717952834988814

In our case:

$$0.038 < 1 - \alpha = 0.95$$

Therefore, we **accept the null hypothesis**. Great!!

### Two sample t-test/Comparison of means

- Instead of drawing an inference on one population through its sample data, we try to compare the means of two populations from which two different sample sets are drawn. It helps us to test whether the population means of the two groups are equal or not.

- This time we will use a T test because population variances are equal, but unknown here. 

- Some of the assumptions on the data before we can use this test are:
  - Populations from which the samples are drawn is assumed to be **normally distributed**
  - Sample data has **continuous values** and **not discrete**
  - Random samples were drawn from the populations (no bias)
  - Data in the two samples must be independent of each other


A psychologist was interested in exploring whether or not male and female college students have different driving behaviors. There were a number of ways that she could quantify driving behaviors. She opted to focus on the fastest speed ever driven by an individual. Therefore, the particular statistical question she framed was as follows:

- Is the mean fastest speed driven by male college students different than the mean fastest speed driven by female college students?

She conducted a survey of a random n = 34 male college students and a random m = 29 female college students. Here is a descriptive summary of the results of her survey:

- Males
samples = 34
Sample mean = 105.5
Sample standard deviation: 20.1

- Females
samples = 29
Sample mean = 90.0
Sample standard deviation:12.2

### Steps in setting up hypothesis Tests:

1. Null Hypothesis
2. Alternate Hypothesis
3. Level of Significance
4. Test Statistic
5. P-value

Step1: Define the null hypothesis 

H0: μ1 = μ2.

Step 2: Define the alternative hypothesis. 

Ha: μ1 != μ2. Since in this case, our null hypothesis would be false on either side of Ha, hence we have a Two tailed Test.

Step 3: Decide a test statistics based on the information available. We will use a t-test.

Step 4: Level of significance: 0.05.

Step 5: Calculate the test statistic.

In [21]:
#males
sample_mean1 = 105.5
sample_std1 = 20.1
n1 = 34
#females
sample_mean2 = 90.9
sample_std2 = 12.2
n2 = 29

pooled_sample_std = math.sqrt(((n1-1)*sample_std1**2 + (n2-1)*sample_std2**2)/(n1+n2-2))
statistic = (sample_mean1-sample_mean2)/(pooled_sample_std*math.sqrt((1/n1)+(1/n2)))
print("T Statistic is: ", statistic)

T Statistic is:  3.4101131776909535


In [25]:
from scipy.stats import t

print("P value is: ", 1- t.cdf(statistic,n1+n2-2))
print("Critical Value of z is: ", t.ppf(0.025, n1+n2-2)) #alpha is 0.05

P value is:  0.0005783712704484634
Critical Value of z is:  -1.9996235841149783


Step 6: Make inferences.

In this case, since the test statistic is greater than the absolute value of "critical value", it is in the rejection region. Hence we reject the null hypothesis.

Also by looking at the p value directly, we can reject the null hypothesis as it is less than 0.05 -> the means of both samples are not equal